r2/service.go

```diff
@@ -509,38 +549,62 @@ type apiResponse struct {
 }
 
 type namespaceJSON struct {
-	ID string `json:"id" validate:"required"`
+	ID                string `json:"id" validate:"required"`
+	ForRuleSetVersion int    `json:"forRuleSetVersion"`
+}
+
+func newNamespaceJSON(nv *rules.NamespaceView) namespaceJSON {
+	return namespaceJSON{
+		ID:                nv.Name,
+		ForRuleSetVersion: nv.ForRuleSetVersion,
+	}
 }
 
 type namespacesJSON struct {
 	Version    int             `json:"version"`
 	Namespaces []namespaceJSON `json:"namespaces"`
 }
 
+func newNamespacesJSON(nss *rules.NamespacesView) namespacesJSON {
+	views := make([]namespaceJSON, len(nss.Namespaces))
+	for i, namespace := range nss.Namespaces {
+		views[i] = newNamespaceJSON(namespace)
+	}
+	return namespacesJSON{
+		Version:    nss.Version,
+		Namespaces: views,
+	}
+}
+
 type mappingRuleJSON struct {
-	ID            string          `json:"id,omitempty"`
-	Name          string          `json:"name" validate:"required"`
-	CutoverMillis int64           `json:"cutoverMillis,omitempty"`
-	Filters       filters         `json:"filter" validate:"required"`
-	Policies      []policy.Policy `json:"policies" validate:"required"`
+	ID                  string          `json:"id,omitempty"`
+	Name                string          `json:"name" validate:"required"`
+	CutoverMillis       int64           `json:"cutoverMillis,omitempty"`
+	Filters             filters         `json:"filter" validate:"required"`
+	Policies            []policy.Policy `json:"policies" validate:"required"`
+	LastUpdatedBy       string          `json:"lastUpdatedBy"`
+	LastUpdatedAtMillis int64           `json:"lastUpdatedAtMillis"`
 }
 
-func (m mappingRuleJSON) mappingRuleView() *rules.MappingRuleView {
+func (m mappingRuleJSON) mappingRuleView(author string) *rules.MappingRuleView {
 	return &rules.MappingRuleView{
-		ID:       m.ID,
-		Name:     m.Name,
-		Filters:  m.Filters,
-		Policies: m.Policies,
+		ID:            m.ID,
+		Name:          m.Name,
+		Filters:       m.Filters,
+		Policies:      m.Policies,
+		LastUpdatedBy: author,
 	}
 }
 
 func newMappingRuleJSON(mrv *rules.MappingRuleView) mappingRuleJSON {
 	return mappingRuleJSON{
-		ID:            mrv.ID,
-		Name:          mrv.Name,
-		Filters:       mrv.Filters,
-		Policies:      mrv.Policies,
-		CutoverMillis: xtime.ToUnixMillis(xtime.FromNanoseconds(mrv.CutoverNanos)),
+		ID:                  mrv.ID,
+		Name:                mrv.Name,
+		Filters:             mrv.Filters,
+		Policies:            mrv.Policies,
+		CutoverMillis:       xtime.ToUnixMillis(xtime.FromNanoseconds(mrv.CutoverNanos)),
```

Comment:
nit: can just divide by `time.Millisecond / time.Nanosecond` to be more efficient.